miner/worker.go

```diff
@@ -572,50 +573,95 @@ func (w *worker) taskLoop() {
 
 		prevParentHash common.Hash
 		prevProfit     *big.Int
+		timer          = time.NewTimer(0)
 	)
 
+	<-timer.C // usual skip first tick
+
 	// interrupt aborts the in-flight sealing task.
 	interrupt := func() {
 		if stopCh != nil {
 			close(stopCh)
 			stopCh = nil
 		}
 	}
+
+	taskWork := map[string]*task{}
+	var defaultTaskPick *task
+
+	handle := func(task *task) {
+		if w.newTaskHook != nil {
+			w.newTaskHook(task)
+		}
+		// Reject duplicate sealing work due to resubmitting.
+		sealHash := w.engine.SealHash(task.block.Header())
+		if sealHash == prev {
+			return
+		}
+
+		taskParentHash := task.block.Header().ParentHash
+		// reject new tasks which don't profit
+		if taskParentHash == prevParentHash &&
+			prevProfit != nil && task.profit.Cmp(prevProfit) < 0 {
+			return
+		}
+		prevParentHash = taskParentHash
+		prevProfit = task.profit
+
+		// Interrupt previous sealing operation
+		interrupt()
+		stopCh, prev = make(chan struct{}), sealHash
+		log.Info("Proposed miner block", "blockNumber", task.block.Number(), "profit", prevProfit, "isFlashbots", task.isFlashbots, "sealhash", sealHash, "parentHash", prevParentHash)
+		if w.skipSealHook != nil && w.skipSealHook(task) {
+			return
+		}
+		w.pendingMu.Lock()
+		w.pendingTasks[sealHash] = task
+		w.pendingMu.Unlock()
+
+		if err := w.engine.Seal(w.chain, task.block, w.resultCh, stopCh); err != nil {
+			log.Warn("Block sealing failed", "err", err)
+		}
+
+	}
+
 	for {
 		select {
-		case task := <-w.taskCh:
-			if w.newTaskHook != nil {
-				w.newTaskHook(task)
-			}
-			// Reject duplicate sealing work due to resubmitting.
-			sealHash := w.engine.SealHash(task.block.Header())
-			if sealHash == prev {
-				continue
-			}
 
-			taskParentHash := task.block.Header().ParentHash
-			// reject new tasks which don't profit
-			if taskParentHash == prevParentHash &&
-				prevProfit != nil && task.profit.Cmp(prevProfit) < 0 {
-				continue
-			}
-			prevParentHash = taskParentHash
-			prevProfit = task.profit
+		case <-timer.C:
+			// timer went off, use first bundle, defaultTaskPick
+			handle(defaultTaskPick)
 
-			// Interrupt previous sealing operation
-			interrupt()
-			stopCh, prev = make(chan struct{}), sealHash
-			log.Info("Proposed miner block", "blockNumber", task.block.Number(), "profit", prevProfit, "isFlashbots", task.isFlashbots, "sealhash", sealHash, "parentHash", prevParentHash)
-			if w.skipSealHook != nil && w.skipSealHook(task) {
-				continue
-			}
-			w.pendingMu.Lock()
-			w.pendingTasks[sealHash] = task
-			w.pendingMu.Unlock()
+		case task := <-w.taskCh:
+			func() {
+				defer func() {
+					// resize the taskWork when it gets huge
+				}()
+
+				num, paren := task.block.Number().Uint64(), task.block.ParentHash()
+				_, plainExists := taskWork[fmt.Sprintf("%d-%v-%v", num, paren, false)]
```

Comment:
I don't think this is unique enough. There are potentially multiple tasks that occur with the same parentHash/blockNumber

Reply:
I was thinking that too - another thing was gonna do- just do the sha256 of the task sans isFlashbots field -

Reply:
That's probably closer, but I do worry that if somehow the block doesn't change (ie. txpool didn't update in the last 3 seconds), then you'd get conflicts. Maybe not a huge deal in practice, as long as the system doesn't crash under this edge case.

Reply:
iirc there are enough timer/heartbeat things in txpool that will give enough churn within 3 seconds -